rand_core/src/seed.rs

```diff
@@ -0,0 +1,48 @@
+/// Trait implemented by types used for seeding PRNG.
+///
+/// This crate provides implementations for `[u8; N]`, `u32`, `u64`, and `u128`.
+pub trait Seed: Sized {
+    /// Create seed from a fallible closure which fills the provided buffer.
+    fn try_from_bytes<E>(fill: impl FnOnce(&mut [u8]) -> Result<(), E>) -> Result<Self, E>;
+
+    /// Create seed from an infallible closure which fills the provided buffer.
+    fn from_bytes(fill: impl FnOnce(&mut [u8])) -> Self {
+        let Ok(seed) = Self::try_from_bytes::<core::convert::Infallible>(|buf| {
+            fill(buf);
+            Ok(())
+        });
+        seed
+    }
+}
+
+impl<const N: usize> Seed for [u8; N] {
+    fn try_from_bytes<E>(fill: impl FnOnce(&mut [u8]) -> Result<(), E>) -> Result<Self, E> {
+        let mut buf = [0u8; N];
+        fill(&mut buf)?;
+        Ok(buf)
+    }
+}
+
+impl Seed for u32 {
+    fn try_from_bytes<E>(fill: impl FnOnce(&mut [u8]) -> Result<(), E>) -> Result<Self, E> {
+        let mut buf = [0u8; 4];
+        fill(&mut buf)?;
+        Ok(u32::from_le_bytes(buf))
+    }
+}
+
+impl Seed for u64 {
+    fn try_from_bytes<E>(fill: impl FnOnce(&mut [u8]) -> Result<(), E>) -> Result<Self, E> {
+        let mut buf = [0u8; 8];
+        fill(&mut buf)?;
+        Ok(u64::from_le_bytes(buf))
+    }
+}
+
+impl Seed for u128 {
```

Comment:
I wonder if we should also implement this trait for `[u32/u64/u128; N]` types. For example, for PCG32 `[u64; 2]` is a "natural" seed, while for ChaCha20 it would be `[u32; 8]`.